Skateboard content expansion pack

itseasytosee opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
(Drafted because working on balance and looking for more rad ideas)

Makes skateboarding a hell of a lot cooler,


Lowers stamina cost for ollieing and grinding.
Makes collisions with other carbons always result in a wipeout no matter your stamina damage, colliding with a carbon will stun and knock them down for a short time (Stunned for the quarter of the length the skater is, and knockdown for have the length) The time is doubled if you collide with them while they are performing a sick grind.

Additionally grinding on top of stuff now has more functionality.
Grinding over a person that is laying down will deal significant damage with a high chance to wound at the cost of the skaters' stamina. For this to happen the person has to be laying down on the table, if they are standing they will simply collide as normal.
You can grind into a disposals bin.
Fixed action buttons always being red
 fixed being able to ollie while unconscious/dead
Grinding over an object that is cuttable with a knife will lead to that item getting instantly cut! This includes cheese.

- [x]  Implement further suggestions for grind functionality
- [x] Fix some known skateboard bugs
- [x] Tweak some stamina cost numbers, maybe stun times
- [x] Make a better changelog
- [x] Record a video where my character doesn't randomly turn invisible because of a memory shortage.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
https://user-images.githubusercontent.com/55666666/150283992-ecf731bc-6c2b-4100-bb38-52966ec05822.mp4

https://user-images.githubusercontent.com/55666666/150379153-ab268574-2c78-48fb-a09f-7d93237d83bc.mp4


https://user-images.githubusercontent.com/55666666/150401922-c89ae889-e1e0-4a18-b60c-c9f9ba6a2d7d.mp4



<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: itseasytosee
expansion: Skateboards have been made considerably more radical
expansion: While grinding on a table with a skateboard, if you grind on top of somebody you are likely to hurt them quite badly 
expansion: Colliding with another person will now always result in a crash, in which the person hit gets knocked over as well as the skater (protip: crash into someone while grind to double the stun time!)
expansion: Colliding with a disposal unit while grinding will result in you getting dunked in! Wack!
fix: fixed some skateboard related bugs, including the action button for ollieing always being red, and being able to ollie while dead
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

itseasytosee committed:
skateboard stuff

cacogen commented:
I wish you could do kickflip or grab trick and it had combos like that shitty MMA garbage we have but this is pretty cool too. I mean if Spyro Dragon can rip off Bam Margera (inventor of skateparks) why not us?

ArcaneMusic commented:
"Who's ready to cut the cake?"
**SPACE TONY HAWK GRINDS IN, OLLIES THROUGH THE AIR KNOCKING A CHILD DOWN, CRYING**
"WooooOOOOOOOOOAH!"
**THE HAWKSTER TURNS THE CAKE INTO 8 SLICES OF PERFECTLY CUT CONFECTIONARY, CHILDREN FLYING LEFT AND RIGHT TO DODGE THE ELDERLY SKATER'S REDICULOUS GRINDS"
"DUUUUUUUUUUUUUDE"
**THE LIVING LEGEND 360 NOSEGRINDS OUT AN AIRLOCK**

timothymtorres commented:
But can you grind inside a disposals half-pipe?

EricZilla commented:
Always wished there was a positive skater quirk that would shorten your stuntime when you collide with walls (or even just airlocks)

itseasytosee committed:
dumped

itseasytosee commented:
Added grinding into disposals

https://user-images.githubusercontent.com/55666666/150379258-c993ba70-6cab-4fb9-92a5-d0fe004811aa.mp4

itseasytosee committed:
fixes + improvement

itseasytosee commented:
Fixed action buttons always being red, fixed being able to ollie while unconscious/dead

https://user-images.githubusercontent.com/55666666/150401556-00a361dd-2da1-4ea2-8ae7-01e563dc9ea7.mp4

Bawhoppen commented:
Good PR, PKP would be proud 👍

itseasytosee committed:
balance tweak

itseasytosee commented:
![trade offer](https://user-images.githubusercontent.com/55666666/150453263-fcd4bb4e-ba8c-4239-9ffe-155c2056a743.jpg)

BadIsMyUsername commented:
> Added grinding into disposals

Can you grind OUT of disposals? Like if someone placed a table at the exit chute, would you continue on the epic grind? Imagine the skate parks possible if so...

Likteer commented:
@itseasytosee

https://user-images.githubusercontent.com/68878861/150698700-314ed30d-156f-46d4-bf79-4c76feaf39a8.mp4

itseasytosee commented:
> @itseasytosee
> 
Thank you, it was inspired

itseasytosee commented:
This should be ready to go for maintainer review, I haven't collected any more ideas on par with the ability to cut cake by skateboarding over it. But I might come back to it and open another PR with more stuff.

as of now :shipit:

tohg7 commented:
> This should be ready to go for maintainer review, I haven't collected any more ideas on par with the ability to cut cake by skateboarding over it. But I might come back to it and open another PR with more stuff.
> 
> as of now :shipit:

One thing you could add is a "Skater" quirk that allows you to spawn with a skateboard in your bag and have reduced self-stun times when crashing into someone + increased stun times for the person you crashed into.

itseasytosee commented:
> > This should be ready to go for maintainer review, I haven't collected any more ideas on par with the ability to cut cake by skateboarding over it. But I might come back to it and open another PR with more stuff.
> > as of now :shipit:
> 
> One thing you could add is an expensive "Pro Skater" quirk that allows you to spawn with a skateboard in your bag and have reduced self-stun times when crashing into someone + increased stun times for the person you crashed into.

I thought about giving you a trait like that if you were wearing a backwards cap but I think that's a PR for the future.

tohg7 commented:
> > > This should be ready to go for maintainer review, I haven't collected any more ideas on par with the ability to cut cake by skateboarding over it. But I might come back to it and open another PR with more stuff.
> > > as of now :shipit:
> > 
> > 
> > One thing you could add is an expensive "Pro Skater" quirk that allows you to spawn with a skateboard in your bag and have reduced self-stun times when crashing into someone + increased stun times for the person you crashed into.
> 
> I thought about giving you a trait like that if you were wearing a backwards cap (or wearing a jetpack by someone's suggestion) but I think that's a PR for the future.

one idea i have for one of your future PRs is bamboo blowguns being reworked to have a shorter progress bar when being fired and are able to move around while the progress bar for firing is ongoing, because currently, the bamboo blowgun is absolute garbage.

tohg7 commented:
> > > This should be ready to go for maintainer review, I haven't collected any more ideas on par with the ability to cut cake by skateboarding over it. But I might come back to it and open another PR with more stuff.
> > > as of now :shipit:
> > 
> > 
> > One thing you could add is an expensive "Pro Skater" quirk that allows you to spawn with a skateboard in your bag and have reduced self-stun times when crashing into someone + increased stun times for the person you crashed into.
> 
> I thought about giving you a trait like that if you were wearing a backwards cap (or wearing a jetpack by someone's suggestion) but I think that's a PR for the future.

I think there should be a crash stun-duration bonus for crashing into someone at high speeds, like from extinguisher boosting into them.

cacogen commented:
rip 
https://www.youtube.com/watch?v=NkRouk5b4ps

itseasytosee commented:
Thanks oranges, ready for maintainer review

Rohesie commented on code/modules/vehicles/vehicle_actions.dm:
```diff
@@ -287,18 +287,17 @@
 
 /datum/action/vehicle/ridden/scooter/skateboard/ollie/Trigger(trigger_flags)
-	if(world.time > next_ollie)
+	if(..())
```
```suggestion
	. = ..()
	if(.)
```

Rohesie commented on code/modules/vehicles/scooter.dm:
```diff
@@ -116,24 +129,43 @@
+				var/total_brute = victim.getBruteLoss()
+				playsound(location, 'sound/items/trayhit2.ogg', 40)
+				if(victim.stat == DEAD && total_brute >= 175)
+					victim.gib() // Beaten and brutalised, your body gives up
```
This is a definite no from me.

itseasytosee replied:
I think I understand where you are coming from, but could you elaborate?

Rohesie replied:
I'd rather not expand on the ways of round removal on a feature like this.
The less balance concerns you introduce, the better.

itseasytosee replied:
Yeah, gibbing is super over the top here in this scenario mostly because it doesn't make sense and I'll remove it.

Generally, I'm all for more round removal and wacky balance but I suppose that's more of a philosophical issue rather than a coding one huh. Balance concerns are the fun part after all.

Rohesie commented on code/modules/vehicles/scooter.dm:
```diff
@@ -116,24 +129,43 @@
+					skater.adjustStaminaLoss(instability)
+					victim.visible_message(span_danger("[victim] straight up gets grinded into the ground by [skater]'s [src]! Radical!"))
+		for(var/obj/item/cut_thing in location)
+			var/cut_holder = new/obj/item/knife(null) // The lore implications is that when a skateboard runs over cheese, it summons a knife from the shadow realm to cut it
+			cut_thing.tool_act(skater, cut_holder, TOOL_KNIFE, instant_proccess = TRUE)
+			qdel(cut_holder) // and then disapears the knife back to where it came from.
```
This is very hacky and requires you to even add an `instant_process` argument and pass it all around.
Don't. Make use of a proc cutting proc instead which the knife's `tool_act()` also uses.

itseasytosee replied:
I'll work on it. I'm not sure if it can be implemented exactly as you suggest.

itseasytosee replied:
I don't know what you mean by "proc cutting proc." 
Are you suggesting I make a new proc that directly sends a `COMSIG_ATOM_TOOL_ACT ` alongside a modification that overrides `time_to_process`? or is that also too hacky?

Rohesie replied:
This is a very good case of componentization. What you are after here is not the knife, but the property it holds: cutting.
That is represented by the `tool_behavior` value of `TOOL_KNIFE`.

The way it works is that the knife's `tool_act()` sends the `COMSIG_ATOM_TOOL_ACT` signal, to which the food and similar things register, and they populate a list of the `processing_recipes` available, triggering `process_recipes()`, which you modified.
The problem, and the reason you are having to pass further arguments, is that they all pass the tool as an argument in order to extract the needed information. So that's why you're having to spawn and delete an item.

The proper way to accomplish what you want is to refactor these procs so that instead of passing the whole item by reference, they pass the necessary information, decoupled from it. Then you'll be able to use the tool act without spawning an object.
This is not necessarily a trivial change, as you'd have to refactor down to the `use_tool()` proc,

Consider if you want to make this change. It would be a really good improvement of the code and I can help you through, but, as said, it is not trivial. And if you do it then you can make your balance changes. Code quality comes first and justifies other changes, to me.

If you are not willing to make the change then just remove the cutting feature.

itseasytosee replied:
Yeah seems easy enough. Cutting feature is non-negotiable and the inspiration of this PR. I will do it.

Rohesie commented on code/modules/vehicles/scooter.dm:
```diff
@@ -82,25 +82,38 @@
+			if(grinding)
+				grinding_mulitipler = 2
+			victim.Knockdown(4 * grinding_mulitipler SECONDS)
+			victim.Paralyze(1 * grinding_mulitipler SECONDS)
```
Let's go slow on the balance changes, please.

itseasytosee replied:
These are less than half of what the skater takes when they collide. This very much improves the feel of skateboarding and opens combat opportunities in which one person skates into someone exposing them to attack from a partner. Seems relatively balanced, and it's not going to shake the meta.
I'd be open to your proposed changes.

itseasytosee replied:
If this PR doesn't change the effectiveness of skateboards as a potential weapon of combat then this PR has failed. I'd be more than willing to accept a balance tag.

Rohesie replied:
The knockdown should be enough for those purposes. No need to `Paralyze()`.

itseasytosee replied:
I've given it some thought and this is a good solution. It will be done.

Rohesie commented on code/game/atoms.dm:
```diff
@@ -1367,7 +1367,7 @@
  * Must return  parent proc ..() in the end if overridden
  */
-/atom/proc/tool_act(mob/living/user, obj/item/tool, tool_type, is_right_clicking)
+/atom/proc/tool_act(mob/living/user, obj/item/tool, tool_type, is_right_clicking, instant_proccess = FALSE)
```
Don't add these new arguments all around. See the other comment.

Rohesie reviewed (changes requested):
Hacky implementation, needs work.

itseasytosee committed:
easy fixes first

itseasytosee committed:
Progress

itseasytosee committed:
redundincies and the like

itseasytosee committed:
and again

itseasytosee commented:
Ok, the code is kicking my ass here. I'll clean up and get rid of the cutting feature. I promise to get back to it before I die.

itseasytosee committed:
cleanup

itseasytosee committed:
sweep sweep sweep

itseasytosee committed:
just the tip

itseasytosee committed:
Finishing touch

Rohesie commented on code/game/atoms.dm:
```diff
@@ -1372,7 +1372,7 @@
 	if(!is_right_clicking) // Left click first for sensibility
 		var/list/processing_recipes = list() //List of recipes that can be mutated by sending the signal
-		signal_result = SEND_SIGNAL(src, COMSIG_ATOM_TOOL_ACT(tool_type), user, tool, processing_recipes)
+		signal_result = SEND_SIGNAL(src, COMSIG_ATOM_TOOL_ACT(tool_type), user, tool, processing_recipes, tool)
```
This change is not intended now, is it?

Rohesie commented on code/game/atoms.dm:
```diff
@@ -1395,7 +1395,7 @@
 				act_result = analyzer_act(user, tool)
 	else
-		signal_result = SEND_SIGNAL(src, COMSIG_ATOM_SECONDARY_TOOL_ACT(tool_type), user, tool)
+		signal_result = SEND_SIGNAL(src, COMSIG_ATOM_SECONDARY_TOOL_ACT(tool_type), user)
```
Nor this one.

Rohesie commented on code/modules/vehicles/vehicle_actions.dm:
```diff
@@ -287,18 +287,18 @@
 
 /datum/action/vehicle/ridden/scooter/skateboard/ollie/Trigger(trigger_flags)
-	if(world.time > next_ollie)
+	. = ..()
+	if(.)
```
Does this completely remove the cooldown?

Rohesie commented on code/modules/vehicles/scooter.dm:
```diff
@@ -82,25 +82,37 @@
+		if((istype(bumped_thing, /obj/machinery/disposal/bin)))
+			rider.Paralyze(8 SECONDS)
+			rider.forceMove(bumped_thing)
+			src.forceMove(bumped_thing)
```
```suggestion
			forceMove(bumped_thing)
```
`src` is implied.

Rohesie commented on code/modules/vehicles/vehicle_actions.dm:
```diff
@@ -287,18 +287,18 @@
 		var/mob/living/rider = owner
 		var/turf/landing_turf = get_step(vehicle.loc, vehicle.dir)
-		rider.adjustStaminaLoss(vehicle.instability*2)
+		rider.adjustStaminaLoss(vehicle.instability* 0.75)
```
Is this buff entirely necessary? Can you explain why?

itseasytosee replied:
The original stamina cost was too high, the base skateboard can only grind over 4 tables maximum before having to take a break and even less if you want to chain skateboard grinds one after another because of the stamina cost of the ollie.
Makes it feel cooler, it should have always been this way.

itseasytosee committed:
overaddition

itseasytosee committed:
mistake

itseasytosee committed:
Apply suggestions from code review

Co-authored-by: Rohesie <rohesie@gmail.com>

Rohesie commented:
Something tells me I'll regret this one.
